Commit to edit targets file, attempting to isolate each target

jholbrook64 opened the pull request:
Update: 
            modified function download_nwis_data to read from all the files in the fetch/out directory and concat them.
            For function download_nwis_site_data, this is called seperatley for each target in the terminal. The goal was to read each target as its full self, and then write code that would allow the targets to skip if the file was already located in the directory.
targets return form as 'file' in order to skip build on repeated runs



![targets2visnetwork](https://user-images.githubusercontent.com/85192408/123145801-0e56c700-d412-11eb-8013-faaddc606696.PNG)

jholbrook64 committed:
Commit to edit targets file, attempting to isolate each target

jholbrook64 committed:
Commit extra method in get_nwis_data.R that only writes
		if the file doesnt exist and and enddate isn't a day
		after  the paramerter for enddate. This is supposed to
		function as a method only accessible between functions in
		get_nwis_data.R and determiens whetyher or not the data
		should be reodownloaded or skipped.

		modified download_nwis_site_data function and the
		download_nwis_data function to only work off established
		filepaths instead of temporary directories.
		download_nwis_data core function is to concat csvs of
		pulled data.

limnoliver commented on _targets.R:
```diff
@@ -2,22 +2,61 @@ library(targets)
+                        # using separate functions for each target
+tar_target(
+  nwis_01427207_data_csv,
+  download_nwis_site_data(filepath = '1_fetch/out/nwis_01466500_data.csv'),
```
The site IDs in the target name and the filepath don't match for several of these targets.

limnoliver commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -1,48 +1,102 @@
   set.seed(Sys.time())
   if (sample(c(T,F,F,F), 1)){
-    stop(site_num, ' has failed due to connection timeout. Try tar_make() again')
+    stop(site_num, ' has failed due to connection timeout. Try tar_make() again')    # does this means it stops at time 1000?
```
This chunk is simulating failure (like a website that times out) by randomly sampling TRUE and FALSE (line 95) - so each time this is run you have a 25% chance of failure.

limnoliver commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -1,48 +1,102 @@
-  # filepaths look something like directory/nwis_01432160_data.csv,
+# 3rd funciton
+download_nwis_site_data <- function(filepath, parameterCd = '00010', startDate="2014-05-01", endDate="2015-05-01")
+  {  # changed this function signature too
```
what does function signature mean?

jholbrook64 replied:
I included dates in the function signature to allow calls to modify the start and end dates, these are set to hard coded values currently, but they could be switched to variables.

limnoliver commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -1,48 +1,102 @@
+    # readNWISdata is from the dataRetrieval package
+    data_out <- readNWISdata(sites=site_num, service="iv", 
+                             parameterCd = parameterCd, startDate = startDate, endDate = endDate)
+    write_csv(data_out, file = filepath)
```
I think this might be cheating a bit ;) since you moved up the write function before the random error generator.

jholbrook64 replied:
this was undone

limnoliver commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -1,48 +1,102 @@
-  # readNWISdata is from the dataRetrieval package
-  data_out <- readNWISdata(sites=site_num, service="iv", 
-                           parameterCd = parameterCd, startDate = startDate, endDate = endDate)
+  if (!dir.exists(file.path("1_fetch/out", paste0('nwis_', site_nums, '_data.csv')))) # this produces warning, but the warning is only about the length of the input, since there is an input vector that is iterated through
```
What are you hoping to do with this directory check?

Now that you've separated the data pulls into different targets for each site, `tar_make` will skip the targets that have been successfully built, and only build the ones that failed previously, haven't been built yet, or are out of date for some reason (a function changed). I don't think you need this check anymore.

jholbrook64 replied:
this control flow was removed but tar_make still attempts to rebuild targets of files still in the out directory currently.

jholbrook64 committed:
commited changes to _targets.r file and the fetch data functions
	in 1_fetch/src/getnwis/data.R

	I need to figure out how to get the targets to skip downloading
	files that are already in the olcal directory.

jholbrook64 committed:
Commits made to return objects instead of filepaths.
	Idea: targets relies off similar objects in order to skip
	process on repeated runs.

	changes: edited control flow in download_nwis_site_data

limnoliver commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -24,25 +57,56 @@ nwis_site_info <- function(fileout, site_data){
-  set.seed(Sys.time())
-  if (sample(c(T,F,F,F), 1)){
-    stop(site_num, ' has failed due to connection timeout. Try tar_make() again')
+  if (dir.exists(file.path("1_fetch/out", paste0('nwis_', site_num, '_data.csv')))) # tar_make wasn't skipping succsesfully, so I included this if/else
```
I think this is just looking for your `filepath` string, so you could use that instead of recreating it here.

limnoliver reviewed (commented):
I was able to successfully run this pipeline after I removed

jholbrook64 committed:
Commit fixes mismathced target names inn _targets.r and removes
	conditions from function download_nwis_site_data

limnoliver commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -1,48 +1,66 @@
+download_nwis_data <- function(site_nums, parameterCd = '00010')  
+{
+  repoFiles <- file.path("1_fetch/out", paste0('nwis_', site_nums, '_data.csv'))
+  browser()
```
You still have a browser in here. Should remove before merging.

jholbrook64 replied:
removed in commit#9

limnoliver commented on _targets.R:
```diff
@@ -2,22 +2,62 @@ library(targets)
 source("1_fetch/src/get_nwis_data.R")
 source("2_process/src/process_and_style.R")
 source("3_visualize/src/plot_timeseries.R")
+tar_option_set(debug = "site_data")
```
You should comment this out for now, too.

jholbrook64 replied:
commented out

limnoliver commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -1,48 +1,66 @@
+                         X_00010_00000 = c(), X00010_00000_cd = c(), tz_cd = c())
+  for (data in repoFiles)
+  {
+    download_nwis_site_data(data, parameterCd = '00010')
```
You're now back to the original problem of this failing a lot because you're calling download_nwis_site_data in each loop, and it has a high chance of failing out of your 5 tries, which kills the entire target. The way you had it before, with separate targets for each of these downloads, was the right approach.

jholbrook64 replied:
fixed this

jholbrook64 committed:
commit - removed download_nwis_site_data function call in data
	download_nwis_data function

jholbrook64 committed:
Commented out debug option:wq

jholbrook64 committed:
changed order of targets in _targets.r

limnoliver commented on _targets.R:
```diff
@@ -47,7 +43,12 @@ tar_target(
+  ),
+tar_target(
+  site_data,
+  download_nwis_data(site_nums = site_nums),       # where this function is now focused on getting the site_data, I think this function needs to be called first
```
I see you've changed the function, but maybe change the function name to reflect what it's doing now? e.g., binding data, not downloading?

jholbrook64 committed:
Commit changes to _targets concat function name, now named
	concat_nwis_data.

jholbrook64 committed:
Commit changes to get_nwis_data.R, where browser() is removed in
	function concat_nwis_data

limnoliver reviewed (approved):
I did not verify it runs, but if it runs on your end, this looks good to me! Nice work.
